### ISSUE TITLE
SPICE - Error in computing filename date string in write_cdf

### ISSUE THREAD

**Issue by tmplummer** [IMAP-Science-Operations-Center/imap_processing#1208]
### Algorithm Description:

While working on SPICE time functions, I discovered that the implementation in `imap_processing/cdf/utils.py::write_cdf` for generating the date string from the zeroth element of the Dataset epoch coordinate is incorrect by 5 seconds due to the occurrences of 5 leap-seconds since the J2000 epoch. This ticket it to evaluate and figure out a solution. Ideas:
- Hard code 5 second offset into calculation. The advantage of doing it this was is that writing the CDF will not require any SPICE kernels be furnished. The disadvantage is that it will be wrong if/when another leap-second occurs.
- Use SPICE. The advantage is it will always be correct assuming the correct leap-second kernel is furnished.

**Comment by vineetbansal** [comment]
I think it's reasonable to package `naif0012.tls` in the `spice` submodule of `imap_processing` so we can use `spiceypy` for this stuff.  This only applies to the `leapseconds` kernel, not in general. It's updated rarely, with the last update in 2017, so it's not too onerous to include future latest leapseconds kernels. Also, my understanding is that the whole concept of leapseconds is being phased out anyway, so this will not turn into an ongoing maintenance headache.

@jtniehof - I'm wondering if you have thoughts on this?

**Comment by tmplummer** [comment]
> It's updated rarely, with the last update in 2017, so it's not too onerous to include future latest leapseconds kernels.

If a leapsecond were to happen in two years who is going to remember to come add the new tls file to the `imap_processing` package and update those hard-coded references to the `naif0012.tls` kernel? I won't.

One other option that I didn't include in the issue description would be to use Astropy. For a long time, we didn't want to add Astropy as a dependency due to it's bloat, but it is a required dependency for `astropy-healpix` so maybe we should use that here instead. I don't know details of how Astropy handles leapseconds, but at least we won't need to remember to make a change.

**Comment by vineetbansal** [comment]
Once dependencies are pinned (as they should be - see my open [issue](https://github.com/IMAP-Science-Operations-Center/sds-data-manager/issues/1252) on this), then using `astropy` for leap seconds doesn't actually eliminate the "remember" problem - it just moves it to having to remember to update `astropy`, whenever it incorporates that knowledge.

Also a leap second update in `astropy` won't come in isolation - it will be bundled with whatever else changed in that `astropy` release, opening up a new can of worms.

**Comment by tmplummer** [comment]
> Once dependencies are pinned (as they should be - see my open [issue](https://github.com/IMAP-Science-Operations-Center/sds-data-manager/issues/1252) on this), then using `astropy` for leap seconds doesn't actually eliminate the "remember" problem - it just moves it to having to remember to update `astropy`, whenever it incorporates that knowledge.
> 
> Also a leap second update in `astropy` won't come in isolation - it will be bundled with whatever else changed in that `astropy` release, opening up a new can of worms.

Nope. Astropy does not need to put out a new release in case of an updated leapseconds table. https://docs.astropy.org/en/latest/api/astropy.utils.iers.LeapSeconds.html

**Comment by tmplummer** [comment]
From another issue that is getting conflated with this one #1296,

>So should we move any imap_processing code calling TTJ2000_EPOCH to fetch the latest kernel using our REST api, furnish them to spice, and then call our functions? (the code in time.py could do this fetching/furnishing itself).

No, we don't really want to be modifying kernel pools inside the processing. The current philosophy  is to furnish all kernels needed up-front and not alter the kernel pool after that. After giving more thought to this, here is my suggestion:
`TTJ2000_EPOCH` is currently only used in one place in the distribution code (`imap_processing.cdf.utils.write_cdf`). For that use case, ttJ2000ns times (epoch) are being used to generate a string with day precision (YYYYMMDD). I think that either truncating or rounding can be done in a way to not need to worry about second precision for this time conversion.

**Comment by jtniehof** [comment]
With the current setup, calculating a datestring is off by five seconds. So if the first record is day N, 00:00:00, the calculated time will be day N, 00:00:05 ... the leap seconds make the actual time "earlier" than the naive time. That's still good for the datestamp in the filename.

The only problem would be if the first timestamp is day N 23:59:55, which seems unlikely.

In general I do think pulling the LSK first, then doing all other calculations including all other kernel loads, makes the most sense, but it doesn't seem necessary for this application.

**Comment by mmshaw** [comment]
@jtniehof @vineetbansal @tmplummer how urgent is this? Should we try for August or can be deferred?

**Comment by tmplummer** [comment]
This can definitely be deferred. Or maybe closed as won't do?
